idn/punycode.go

```diff
@@ -52,6 +59,12 @@ func ToPunycode(s string) string {
 
 // FromPunycode returns unicode domain name from provided punycode string.
 func FromPunycode(s string) string {
+	// Early check to see if decoding is needed.
+	// This will prevent making heap allocations when not needed.
+	if !strings.Contains(s, _PREFIX) {
```

Comment:
[ Quoting <notifications@github.com> in "Re: [dns] Do simple prechecks to av..." ]
>> @@ -52,6 +59,12 @@ func ToPunycode(s string) string {
>>
>>  // FromPunycode returns unicode domain name from provided punycode string.
>>  func FromPunycode(s string) string {
>> +	// Early check to see if decoding is needed.
>> +	// This will prevent making heap allocations when not needed.
>> +	if !strings.Contains(s, _PREFIX) {
>
>But doing a split already allocates memory on the heap which is what we want to
>prevent in most cases with this check. That's why I'm doing the Contains which
>might return true for some cases where we don't need to decode anything. But
>that is ok cause then the other decode code just won't do anything. And it's is
>only in a very small number of cases.

True, but it is very wrong in those cases.

>I guess I could write a quick loop with `dns.NextLabel` to check with
>`strings.HasPrefix` for `_PREFIX`. This would be more correct but it would also
>be slower.
>
>Which option do you prefer?

I think the loop with NextLabel is the right thing to do.

/Miek

--
Miek Gieben